### ISSUE TITLE
Vulkan: Documentation and/or buildsystem changes needed for MoltenVK

### ISSUE THREAD

**Issue by anissen** [godotengine/godot#36213]
**Godot version:**
Latest master (415a901b68) built as `release_debug`

**OS/device including version:**
MacBook Air (13-inch, Mid 2012) w. Intel HD Graphics 4000 1536 MB
macOS Catalina 10.15.2 (19C57)

**Issue description:**
After the Vulkan-branch is merged, Godot fails to run in both Vulkan and GLES2 mode.

Vulkan:
```
./bin/godot.osx.opt.tools.64 --video-driver Vulkan
arguments
0: ./bin/godot.osx.opt.tools.64
1: --video-driver
2: Vulkan
Current path: /Users/nissen/code/godot/godot-source
Godot Engine v4.0.dev.custom_build.415a901b6 - https://godotengine.org
ERROR: No surface extension found, is a driver installed?
   at: _initialize_extensions (drivers/vulkan/vulkan_context.cpp:255)
ERROR: Method/Function Failed, returning: m_value
   at: initialize (platform/osx/os_osx.mm:1596)
```

GLES2:
```
./bin/godot.osx.opt.tools.64 --video-driver GLES2 
arguments
0: ./bin/godot.osx.opt.tools.64
1: --video-driver
2: GLES2
Current path: /Users/nissen/code/godot/godot-source
Godot Engine v4.0.dev.custom_build.415a901b6 - https://godotengine.org
handle_crash: Program crashed with signal 11
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[1] 1   libsystem_platform.dylib            0x00007fff6408242d _sigtramp + 29
[2] 2   ???                                 0x00007ffee4220840 0x0 + 140732725856320
```

**Steps to reproduce:**
N/A

**Minimal reproduction project:**
N/A

**Comment by akien-mga** [comment]
See https://godotengine.org/article/headsup-vulkan-merged-master-branch-unstable

There's no GLES2 backend right now, so it's normal that it doesn't work.

As for Vulkan, it seems it's not working on your system. Not sure if MoltenVK has any specific setup requirements @bruvzg?

**Comment by timothyqiu** [comment]
If MoltenVK is installed via homebrew (`brew install molten-vk`), you can put a `MoltenVK_icd.json` at `~/.local/share/vulkan/icd.d/MoltenVK_icd.json` with the content:

```json
{
    "file_format_version" : "1.0.0",
    "ICD": {
        "library_path": "/usr/local/opt/molten-vk/lib/libMoltenVK.dylib",
        "api_version" : "1.0.0"
    }
}
```

Then everything is done.  Alternatively, you can setup the `VK_ICD_FILENAMES` environment variable to point to `/usr/local/opt/molten-vk/share/vulkan/icd.d/MoltenVK_icd.json`.

Similar steps go for a standalone SDK download, just change the path accordingly.

<https://vulkan.lunarg.com/doc/sdk/1.1.108.0/mac/getting_started.html>

**Comment by bruvzg** [comment]
- LunarG Vulkan SDK have installer script (`./install_vulkan.py`, installs to the `/usr/local/...`).
- You can copy `libMoltenVK.dylib` to the `Frameworks` subfolder of the Godot app bundle (but this won't work when running from terminal).
- You can use `use_static_mvk=yes` build flag (SDK should be installed or added to library search path manually) to link it statically.

**Comment by akien-mga** [comment]
So I guess this needs documentation for users who build from source for macOS/iOS.

Or should we consider bundling MoltenVK in `thirdparty` and building it statically out of the box, so that there's no requirement on the user side?

> You can use `use_static_mvk=yes` build flag (SDK should be installed or added to library search path manually) to link it statically.

I guess for official builds that will be the best solution.

**Comment by bruvzg** [comment]
> Or should we consider bundling MoltenVK in thirdparty and building it statically out of the box, so that there's no requirement on the user side?

It's quite big, build is complex and will take more time than building Godot itself. Also, statically linked MoltenVK doesn't support validation layers, or overriding driver via `VK_ICD_FILENAMES` variable.

**Comment by anissen** [comment]
Turns out this was a documentation issue. I didn't think I had to install Vulkan drivers (because Vulkan is not running on my device, only Metal which is natively supported).

I tried installing with brew and setting the `VK_ICD_FILENAMES` environment variable, as @timothyqiu suggests, but it didn't work for me. It did throw me a different error message though:

```
./bin/godot.osx.opt.tools.64 --driver Vulkan
arguments
0: ./bin/godot.osx.opt.tools.64
1: --driver
2: Vulkan
Current path: /Users/nissen/code/godot/godot-source
Godot Engine v4.0.dev.custom_build.415a901b6 - https://godotengine.org
ERROR: Condition "err" is true. returned: ERR_CANT_CREATE
   at: _initialize_extensions (drivers/vulkan/vulkan_context.cpp:217)
ERROR: Method/Function Failed, returning: m_value
   at: initialize (platform/osx/os_osx.mm:1596)
```

I got Godot to work by simply installing the latest official Vulkan SDK for Mac (https://vulkan.lunarg.com/sdk/home) and running `./install_vulkan.py` as per the Vulkan documentation. 🎉

**The lesson is: Install the Vulkan SDK before trying to run Godot on macOS with the Vulkan video driver.**

Thanks for your help! 🍰

**Comment by anissen** [comment]
Feel free to reopen if you want to use this issue to track the need for Vulkan documentation and/or changes to the build system.

**Comment by akien-mga** [comment]
Yes, that was the idea from the issue renaming. We need to document this and figure out how to handle official builds to minimize effort in the user end (we can't really have macOS players install the Vulkan SDK to play Godot games).

**Comment by mclark4386** [comment]
Would love to help with this however I can^^ Just ran into it and would like to make it better for everyone^^

**Comment by hieunc278** [comment]
> Turns out this was a documentation issue. I didn't think I had to install Vulkan drivers (because Vulkan is not running on my device, only Metal which is natively supported).
> 
> I tried installing with brew and setting the `VK_ICD_FILENAMES` environment variable, as @timothyqiu suggests, but it didn't work for me. It did throw me a different error message though:
> 
> ```
> ./bin/godot.osx.opt.tools.64 --driver Vulkan
> arguments
> 0: ./bin/godot.osx.opt.tools.64
> 1: --driver
> 2: Vulkan
> Current path: /Users/nissen/code/godot/godot-source
> Godot Engine v4.0.dev.custom_build.415a901b6 - https://godotengine.org
> ERROR: Condition "err" is true. returned: ERR_CANT_CREATE
>    at: _initialize_extensions (drivers/vulkan/vulkan_context.cpp:217)
> ERROR: Method/Function Failed, returning: m_value
>    at: initialize (platform/osx/os_osx.mm:1596)
> ```
> 
> I got Godot to work by simply installing the latest official Vulkan SDK for Mac (https://vulkan.lunarg.com/sdk/home) and running `./install_vulkan.py` as per the Vulkan documentation. 🎉
> 
> **The lesson is: Install the Vulkan SDK before trying to run Godot on macOS with the Vulkan video driver.**
> 
> Thanks for your help! 🍰

Hello @anissen,
Thanks for your info. 

I have tried as you said, that I installed Vulkan SDK then run Godot, however, it still have another error that Godot could not find libMoltenVK.dylib Image.

ERROR: GENERAL - Message Id Number: 0 | Message Id Name: Loader Message
	dlopen(/Users/mac/Documents/work/godot/Godot.app/Contents/Resources/vulkan/icd.d/../../../Frameworks/libMoltenVK.dylib, 5): image not found
	Objects - 1
		Object[0] - VK_OBJECT_TYPE_INSTANCE, Handle 140549449556992
   at: _debug_messenger_callback (drivers/vulkan/vulkan_context.cpp:137)

Did you have this error ? Or do you have any idea about that.

I have copy libMoltenVK.dylib into Users/mac/Documents/work/godot/Godot.app/Contents directory, but it still occur this error.

Best regards,

**Comment by sjml** [comment]
Hi -- I'm preparing a PR to fix the issues around codesigning Mac Mono exports (#43732) and part of that will include a change to the template build script as well. Have we had any clarity on whether or how best to include `libMoltenVK.dylib` in exports, or for that matter, in the editor distributions? 

I just figured while I'm in there it could be easy to make the change if there's agreement on it. At the moment, though, I'm just copying in the built `libMoltenVK.dylib` from homebrew's installation, which I'm guessing will not be the preferred source of it.

**Comment by invisiblesun** [comment]
Greetings, I am making a try at 4.0 on a mac book air and installed of Vulkan (latest as of today) per the above link (https://vulkan.lunarg.com/sdk/home). 

Godot bombs out complaining that metal 2.2 isn't available (I am working from Mojave) . I may need to revert back to 3.2 and defer using sqlite (or try the options available for 3.2), but I want to give 4.0 a try. 

the errors I see are:
Using Vulkan Device #0: INTEL - Intel(R) HD Graphics 5300
[mvk-error] VK_ERROR_INVALID_SHADER_NV: Unable to convert SPIR-V to MSL:
MSL conversion error: thread_index_in_simdgroup requires Metal 2.2 in fragment shaders.


ERROR:  - Message Id Number: 0 | Message Id Name: 
	VK_ERROR_INVALID_SHADER_NV: Unable to convert SPIR-V to MSL:
MSL conversion error: thread_index_in_simdgroup requires Metal 2.2 in fragment shaders.


	Objects - 1
		Object[0] - VK_OBJECT_TYPE_SHADER_MODULE, Handle 140348414013216
   at: _debug_messenger_callback (drivers/vulkan/vulkan_context.cpp:155)
[mvk-error] VK_ERROR_INVALID_SHADER_NV: Fragment shader function could not be compiled into pipeline. See previous logged error.
ERROR:  - Message Id Number: 0 | Message Id Name: 
	VK_ERROR_INVALID_SHADER_NV: Fragment shader function could not be compiled into pipeline. See previous logged error.
	Objects - 1
		Object[0] - VK_OBJECT_TYPE_PIPELINE, Handle 140348445257216
   at: _debug_messenger_callback (drivers/vulkan/vulkan_context.cpp:155)
ERROR: vkCreateGraphicsPipelines failed with error -1000012000.
   at: render_pipeline_create (drivers/vulkan/rendering_device_vulkan.cpp:5695)
[mvk-error] VK_ERROR_INVALID_SHADER_NV: Unable to convert SPIR-V to MSL:
MSL conversion error: thread_index_in_simdgroup requires Metal 2.2 in fragment shaders.


ERROR:  - Message Id Number: 0 | Message Id Name: 
	VK_ERROR_INVALID_SHADER_NV: Unable to convert SPIR-V to MSL:
MSL conversion error: thread_index_in_simdgroup requires Metal 2.2 in fragment shaders.


	Objects - 1
		Object[0] - VK_OBJECT_TYPE_SHADER_MODULE, Handle 140348414013216
   at: _debug_messenger_callback (drivers/vulkan/vulkan_context.cpp:155)
[mvk-error] VK_ERROR_INVALID_SHADER_NV: Fragment shader function could not be compiled into pipeline. See previous logged error.
ERROR:  - Message Id Number: 0 | Message Id Name: 
	VK_ERROR_INVALID_SHADER_NV: Fragment shader function could not be compiled into pipeline. See previous logged error.
	Objects - 1
		Object[0] - VK_OBJECT_TYPE_PIPELINE, Handle 140348445257216
   at: _debug_messenger_callback (drivers/vulkan/vulkan_context.cpp:155)
ERROR: vkCreateGraphicsPipelines failed with error -1000012000.
   at: render_pipeline_create (drivers/vulkan/rendering_device_vulkan.cpp:5695)
 
Registered camera FaceTime Camera with id 1 position 0 at index 0
ERROR: Condition "default_certs != nullptr" is true.
   at: load_default_certificates (modules/mbedtls/crypto_mbedtls.cpp:301)

**Comment by akien-mga** [comment]
I think the documentation is up-to-date now on how to build Godot and include MoltenVK: https://docs.godotengine.org/en/latest/development/compiling/compiling_for_osx.html

I'll close this, if there are still things which are unclear it's probably best discussed on a new issue in https://github.com/godotengine/godot-docs (unless it's engine-specific issues).
